File: functions/bookings/booking_settlement_math.js
/* eslint-disable require-jsdoc */
// 檔案名稱：functions/bookings/booking_settlement_math.js
// 功能說明：住宿／安親共用結算：最終應收、實收淨額、待補／待退。

function toInt(value, fallback = 0) {
  if (typeof value === "number" && Number.isFinite(value)) {
    return Math.round(value);
  }
  const parsed = Number.parseInt(String(value == null ? "" : value).trim(), 10);
  return Number.isFinite(parsed) ? parsed : fallback;
}

function extraChargeSum(booking) {
  const data = booking || {};
  let sum = 0;
  if (Array.isArray(data.extraCharges)) {
    data.extraCharges.forEach((item) => {
      if (item && typeof item === "object") {
        sum += toInt(item.amount, 0);
      }
    });
  }
  if (sum <= 0) {
    sum = toInt(data.extraFee, 0);
  }
  return sum;
}

function quotedTotal(booking) {
  const data = booking || {};
  const stored = toInt(
      data.quotedTotalPrice || data.originalTotal || data.estimateTotalPrice,
      0,
  );
  if (stored > 0) {
    return stored;
  }
  const total = toInt(
      data.totalPayableAmount || data.totalPrice || data.totalAmount,
      0,
  );
  const extras = extraChargeSum(data) + toInt(data.manualAdjust, 0);
  return Math.max(0, total - extras);
}

function expectedTotal(booking, manualAdjustOverride) {
  const data = booking || {};
  const quoted = quotedTotal(data);
  const extra = extraChargeSum(data);
  const overtime = toInt(data.overtimeAmount, 0);
  const manual = manualAdjustOverride == null ?
    toInt(data.manualAdjust, 0) : toInt(manualAdjustOverride, 0);
  return Math.max(0, quoted + extra + overtime + manual);
}

function paidAmount(booking) {
  const data = booking || {};
  const paid = toInt(data.paidAmount, 0);
  if (paid > 0) {
    return paid;
  }
  const paymentPaid = toInt(data.paymentPaidAmount, 0);
  if (paymentPaid > 0) {
    return paymentPaid;
  }
  if (data.depositPaid === true ||
      String(data.depositStatus || "").trim() === "confirmed") {
    return toInt(data.depositAmount, 0);
  }
  return 0;
}

function isSettlementConfirmed(booking) {
  const data = booking || {};
  return data.settlementConfirmed === true ||
    data.settledAt != null ||
    data.finalSettlementAmount != null;
}

function isSettlementLocked(booking) {
  return (booking || {}).settlementLocked === true;
}

function canLock(booking) {
  if (!isSettlementConfirmed(booking) || isSettlementLocked(booking)) {
    return false;
  }
  return remainingDue(booking) <= 0 && refundDue(booking) <= 0;
}

function refundedAmount(booking) {
  return toInt((booking || {}).refundAmount, 0);
}

function netCollected(booking) {
  return Math.max(0, paidAmount(booking) - refundedAmount(booking));
}

function remainingDue(booking, manualAdjustOverride) {
  return Math.max(
      0,
      expectedTotal(booking, manualAdjustOverride) - netCollected(booking),
  );
}

function refundDue(booking, manualAdjustOverride) {
  return Math.max(
      0,
      netCollected(booking) - expectedTotal(booking, manualAdjustOverride),
  );
}

function paymentStatusOf(expected, net) {
  const remain = Math.max(0, expected - net);
  const refund = Math.max(0, net - expected);
  if (remain > 0) {
    return net > 0 ? "partial" : "unpaid";
  }
  if (refund > 0) {
    return "refund_pending";
  }
  return expected > 0 || net > 0 ? "paid" : "unpaid";
}

function settlementFields(booking, manualAdjustOverride) {
  const expected = expectedTotal(booking, manualAdjustOverride);
  const paid = paidAmount(booking);
  const refunded = refundedAmount(booking);
  const net = Math.max(0, paid - refunded);
  return {
    quotedTotalPrice: quotedTotal(booking),
    extraChargeSum: extraChargeSum(booking),
    expectedTotal: expected,
    totalPayableAmount: expected,
    totalPrice: expected,
    totalAmount: expected,
    paidAmount: paid,
    refundAmount: refunded,
    remainingAmount: Math.max(0, expected - net),
    refundDueAmount: Math.max(0, net - expected),
    paymentStatus: paymentStatusOf(expected, net),
  };
}

module.exports = {
  toInt,
  extraChargeSum,
  quotedTotal,
  expectedTotal,
  paidAmount,
  refundedAmount,
  netCollected,
  remainingDue,
  refundDue,
  paymentStatusOf,
  settlementFields,
  isSettlementConfirmed,
  isSettlementLocked,
  canLock,
};

File: lib/core/services/booking_settlement_math.dart
// 檔案名稱：lib/core/services/booking_settlement_math.dart
// 功能說明：住宿／安親結算差額：最終應收、實收淨額、待補／待退。

import 'package:petnest_saas/core/models/booking_kind.dart';
import 'package:petnest_saas/core/utils/safe_parse.dart';

class BookingSettlementMath {
  BookingSettlementMath._();

  static int extraChargeSum(Map<String, dynamic> data) {
    int sum = 0;
    final Object? raw = data['extraCharges'];
    if (raw is List) {
      for (final Object? item in raw) {
        if (item is Map) {
          sum += SafeParse.parseMoney(item['amount']);
        }
      }
    }
    if (sum <= 0) {
      sum = SafeParse.parseMoney(data['extraFee']);
    }
    return sum;
  }

  static int quotedTotal(Map<String, dynamic> data) {
    final int stored = SafeParse.parseMoney(
      data['quotedTotalPrice'] ??
          data['originalTotal'] ??
          data['estimateTotalPrice'],
    );
    if (stored > 0) {
      return stored;
    }
    final int total = SafeParse.parseMoney(
      data['totalPayableAmount'] ?? data['totalPrice'] ?? data['totalAmount'],
    );
    final int extras =
        extraChargeSum(data) + SafeParse.parseMoney(data['manualAdjust']);
    final int quoted = total - extras;
    return quoted < 0 ? 0 : quoted;
  }

  static int expectedTotal({
    required Map<String, dynamic> data,
    int? manualAdjustOverride,
  }) {
    final int quoted = quotedTotal(data);
    final int extra = extraChargeSum(data);
    final int overtime = SafeParse.parseMoney(data['overtimeAmount']);
    final int manual =
        manualAdjustOverride ?? SafeParse.parseMoney(data['manualAdjust']);
    final int expected = quoted + extra + overtime + manual;
    return expected < 0 ? 0 : expected;
  }

  static int paidAmount(Map<String, dynamic> data) {
    final int stored = SafeParse.parseMoney(data['paidAmount']);
    if (stored > 0) {
      return stored;
    }
    final int paymentPaid = SafeParse.parseMoney(data['paymentPaidAmount']);
    if (paymentPaid > 0) {
      return paymentPaid;
    }
    final bool depositConfirmed =
        data['depositPaid'] == true ||
        (data['depositStatus'] ?? '').toString() == 'confirmed';
    if (depositConfirmed) {
      return SafeParse.parseMoney(data['depositAmount']);
    }
    return 0;
  }

  static bool isSettlementConfirmed(Map<String, dynamic> data) {
    return data['settlementConfirmed'] == true ||
        data['settledAt'] != null ||
        data['finalSettlementAmount'] != null;
  }

  static bool isSettlementLocked(Map<String, dynamic> data) {
    return data['settlementLocked'] == true;
  }

  static bool canLock(Map<String, dynamic> data) {
    if (!isSettlementConfirmed(data) || isSettlementLocked(data)) {
      return false;
    }
    return remainingDue(data: data) <= 0 && refundDue(data: data) <= 0;
  }

  static int remainingDue({
    required Map<String, dynamic> data,
    int? manualAdjustOverride,
  }) {
    final int delta = balanceDelta(
      data: data,
      manualAdjustOverride: manualAdjustOverride,
    );
    return delta > 0 ? delta : 0;
  }

  static int refundDue({
    required Map<String, dynamic> data,
    int? manualAdjustOverride,
  }) {
    final int delta = balanceDelta(
      data: data,
      manualAdjustOverride: manualAdjustOverride,
    );
    return delta < 0 ? -delta : 0;
  }

  static int refundedAmount(Map<String, dynamic> data) {
    return SafeParse.parseMoney(data['refundAmount']);
  }

  static int netCollected(Map<String, dynamic> data) {
    final int net = paidAmount(data) - refundedAmount(data);
    return net < 0 ? 0 : net;
  }

  static int balanceDelta({
    required Map<String, dynamic> data,
    int? manualAdjustOverride,
  }) {
    return expectedTotal(
          data: data,
          manualAdjustOverride: manualAdjustOverride,
        ) -
        netCollected(data);
  }

  static bool isDaycare(Map<String, dynamic> data) {
    return BookingKind.isDaycare(data);
  }

  /// 安親須已結算且款項結清。住宿結算後同樣看差額；尚未結算的舊住宿仍看 status。
  static bool isOrderComplete(Map<String, dynamic> data) {
    final String status = (data['status'] ?? '').toString();
    if (status == 'cancelled' || status == 'no_show') {
      return false;
    }
    if (isDaycare(data)) {
      if (!isSettlementConfirmed(data)) {
        return false;
      }
      return remainingDue(data: data) <= 0 && refundDue(data: data) <= 0;
    }
    if (isSettlementConfirmed(data)) {
      return remainingDue(data: data) <= 0 && refundDue(data: data) <= 0;
    }
    return status == 'completed';
  }

  static bool isDaycareAwaitingClear(Map<String, dynamic> data) {
    return isDaycare(data) &&
        isSettlementConfirmed(data) &&
        !isOrderComplete(data);
  }

  static bool isStayAwaitingClear(Map<String, dynamic> data) {
    return !isDaycare(data) &&
        isSettlementConfirmed(data) &&
        !isOrderComplete(data);
  }

  static bool isAwaitingClear(Map<String, dynamic> data) {
    return isDaycareAwaitingClear(data) || isStayAwaitingClear(data);
  }
}
